bears/general/FilenameBear.py

```diff
@@ -74,12 +83,44 @@ def run(self, filename, file,
         if ignore_uppercase_filenames and filename_without_extension.isupper():
             return
 
-        if new_name != filename_without_extension:
-            diff = Diff(file, rename=os.path.join(head, new_name + extension))
+        if (new_name != filename_without_extension or
+            not filename_without_extension.startswith(filename_prefix) or
+                not filename_without_extension.endswith(filename_suffix)):
+            correct_prefix_name = (new_name
+                                   if
+                                   filename_without_extension.startswith(
+                                                               filename_prefix)
+                                   else
+                                   filename_prefix + new_name)
```

Comment:
Should be a new issue, but I think I got a smart idea :P

Instead of just prepending the prefix if we don't match it in the beginning, we try to match it as long as possible. I.e. consider this:
A file is called `hello_world.py`. We want it to have the prefix `python_` (whyever...). Our patch would suggest to correct it to `python_hello_world`. This is fine. But now consider a filename with a typo:
`pythonhello_world2.py`. Our bear would propose now `python_pythonhello_world2.py`. A human would see that we probably meant to type `python_hello_world2.py`. Same with this file: `ython_hello_world3.py` is probably meant to be `python_hello_world3.py`.

Like said this is imo a new issue, but we could try to perform some best-suiting match of the prefix (same idea would also apply for suffixes) and suggest a smarter patch :)